# Data Cleaning

In [ ]:
import pandas as pd

### Importing the RAW dataset for cleaning

In [69]:
reviews_raw = pd.read_csv("C:\\Users\\saite\\OneDrive - University of Central Missouri\\Documents\\MS in CS\\Data Analyst\\Projects\\Employee Rentention Analysis\\Datasets\\Glassdoor Reviews Cleaned.csv")

In [70]:
reviews_raw.head()

Employeer ID Employeer Name Employement Status              job title  \
0       5224839  JPMorganChase            REGULAR  Private Client Banker   
1       5224839  JPMorganChase            REGULAR       Senior Associate   
2       5224839  JPMorganChase            REGULAR        Personal Banker   
3       5224839  JPMorganChase            REGULAR  Software Engineer III   
4       5224839  JPMorganChase            REGULAR                    NaN   

   length of employement         location  isCurrent Job  \
0                      9  Los Angeles, CA           True   
1                      6     New York, NY           True   
2                      4      Chicago, IL           True   
3                      2      Seattle, WA           True   
4                      1              NaN          False   

   ratingCarrerOpportunities ratingBusinessoutlook   ratingCEO  ...  \
0                          4              POSITIVE     APPROVE  ...   
1                          5              POSITIVE     APPROVE  ...   
2                          2                   NaN     APPROVE  ...   
3                          2              NEGATIVE  DISAPPROVE  ...   
4                          3                   NaN         NaN  ...   

   rating RecommendtoFriend  ratingSeniorleadership  rating worklifebalance  \
0                  POSITIVE                     4.0                     5.0   
1                  POSITIVE                     5.0                     5.0   
2                  NEGATIVE                     2.0                     3.0   
3                  NEGATIVE                     1.0                     4.0   
4                  NEGATIVE                     1.0                     2.0   

     reviewdatetime countHelpful  countNothelpful  \
0  18-05-2025 23:05            0                0   
1  31-05-2025 09:15            0                0   
2  23-05-2025 21:00            0                0   
3  13-05-2025 08:11            0                0   
4  23-05-2025 07:39            0                0   

                                             summary  \
0       Great Company, Location &amp; Team Dependent   
1                                 Good place to work   
2  Great Brand and Resources, But Support Varies ...   
3  Terrible developer experience in a low trust e...   
4                                      Disappointing   

                                              Advice  \
0  Increase base pay for private client bankers. ...   
1                                                NaN   
2  Advice to Management:\nInvest more in developi...   
3  Ditch the LLMs, focus on process efficiency, f...   
4  360 annual reviews for management \nExit inter...   

                                                Cons  \
0  Dependent on team and location. Lots of widget...   
1                                Pay is on lower end   
2  •\tThe employee experience depends heavily on ...   
3  Developer experience at this company is catast...   
4  Poor management who have been granted permissi...   

                                                pros  
0  Great place to learn the industry. If you're i...  
1       Friendly people, collaborative, good culture  
2  •\tThe company offers a well-known name and so...  
3                    Interesting problems to work on  
4  Compensation  \nIt looks good on your resume  ...  

[5 rows x 24 columns]

### Lookup if the dataset consists of any null values

In [71]:
reviews_raw.isnull().sum()

Employeer ID                         0
Employeer Name                       0
Employement Status                 148
job title                          915
length of employement                0
location                          1684
isCurrent Job                        0
ratingCarrerOpportunities            0
ratingBusinessoutlook             4970
ratingCEO                         4968
ratingCompensationand Benefits       0
ratingCulutreand Values              0
ratingDiversity and Inclusion     3178
rating Overall                    3178
rating RecommendtoFriend          6383
ratingSeniorleadership            3178
rating worklifebalance            3178
reviewdatetime                    3178
countHelpful                         0
countNothelpful                      0
summary                           3217
Advice                            7644
Cons                                 0
pros                                 0
dtype: int64

#### Lets look how much percentage of data is missing in each column

In [72]:
round(reviews_raw.isnull().mean() * 100,2)

Employeer ID                       0.00
Employeer Name                     0.00
Employement Status                 1.24
job title                          7.67
length of employement              0.00
location                          14.12
isCurrent Job                      0.00
ratingCarrerOpportunities          0.00
ratingBusinessoutlook             41.68
ratingCEO                         41.66
ratingCompensationand Benefits     0.00
ratingCulutreand Values            0.00
ratingDiversity and Inclusion     26.65
rating Overall                    26.65
rating RecommendtoFriend          53.53
ratingSeniorleadership            26.65
rating worklifebalance            26.65
reviewdatetime                    26.65
countHelpful                       0.00
countNothelpful                    0.00
summary                           26.98
Advice                            64.10
Cons                               0.00
pros                               0.00
dtype: float64

In [73]:
reviews_raw_copy = reviews_raw.copy()

#### Based on the percentage of missing data, the higher % of missing values are in Advice, RecommendtoFriend columns we will drop these columns which are not much important for analysis

In [74]:
reviews_raw_copy.drop(columns=['Advice', 'rating RecommendtoFriend'],inplace=True)

#### For columns like ratingBusinessoutlook, ratingCEO we will keep these columns.
#### We will drop null values from columns like Employement Status, job title, location, Diversity and inclusion, Overall, SeniorLeadership, worklifebalance, summary

In [75]:
reviews_raw_copy.dropna(subset=['Employement Status','job title','location','ratingDiversity and Inclusion', 'rating Overall', 'ratingSeniorleadership', 'rating worklifebalance', 'summary'],inplace=True)

In [76]:
reviews_raw_copy.shape

(7160, 22)

#### So now the dataset shape reduced from (11925, 24) to (7160, 22) which is 39.96 ~ 40% of data has been removed.

In [77]:
# How much percentage of data has been removed after cleaning
removed_data_percentage = round((1 - (reviews_raw_copy.shape[0] / reviews_raw.shape[0])) * 100, 2)
removed_data_percentage

39.96

## Now updating some columns data which can be used to analysis which are like date

In [78]:
reviews_raw_copy['reviewdatetime'] = pd.to_datetime(reviews_raw_copy['reviewdatetime'])

C:\Users\saite\AppData\Local\Temp\ipykernel_36060\292165798.py:1: UserWarning: Parsing dates in %d-%m-%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  reviews_raw_copy['reviewdatetime'] = pd.to_datetime(reviews_raw_copy['reviewdatetime'])


## We will divide the datetime into two columns : Date and Time and then remove reviewDatetime column

In [79]:
reviews_raw_copy['reviewdate'] = reviews_raw_copy['reviewdatetime'].dt.date
reviews_raw_copy['reviewtime'] = reviews_raw_copy['reviewdatetime'].dt.time

In [80]:
reviews_raw_copy.drop(columns=['reviewdatetime'],inplace=True)

### Converting Data Types of ratingDiversity and Inclusion, rating Overall, ratingSeniorleadership, rating worklifebalance to int

In [81]:
reviews_raw_copy['ratingDiversity and Inclusion'] = reviews_raw_copy['ratingDiversity and Inclusion'].astype(int)
reviews_raw_copy['rating Overall'] = reviews_raw_copy['rating Overall'].astype(int)
reviews_raw_copy['ratingSeniorleadership'] = reviews_raw_copy['ratingSeniorleadership'].astype(int)
reviews_raw_copy['rating worklifebalance'] = reviews_raw_copy['rating worklifebalance'].astype(int)

In [82]:
reviews_raw_copy

Employeer ID   Employeer Name Employement Status  \
0           5224839    JPMorganChase            REGULAR   
1           5224839    JPMorganChase            REGULAR   
2           5224839    JPMorganChase            REGULAR   
3           5224839    JPMorganChase            REGULAR   
5           5224839    JPMorganChase            REGULAR   
...             ...              ...                ...   
11919           683  United Airlines            REGULAR   
11920           683  United Airlines            REGULAR   
11921           683  United Airlines          PART_TIME   
11922           683  United Airlines            REGULAR   
11923           683  United Airlines            REGULAR   

                                          job title  length of employement  \
0                             Private Client Banker                      9   
1                                  Senior Associate                      6   
2                                   Personal Banker                      4   
3                             Software Engineer III                      2   
5                                   Process Manager                     20   
...                                             ...                    ...   
11919  Reservation Sales and Service Representative                      1   
11920                              Flight Attendant                      0   
11921                         Ramp Service Employee                      4   
11922                            Operations Manager                      0   
11923                              Flight Attendant                      0   

              location  isCurrent Job  ratingCarrerOpportunities  \
0      Los Angeles, CA           True                          4   
1         New York, NY           True                          5   
2          Chicago, IL           True                          2   
3          Seattle, WA           True                          2   
5         Columbus, OH           True                          3   
...                ...            ...                        ...   
11919      Chicago, IL          False                          4   
11920  Los Angeles, CA           True                          0   
11921    Barrigada, GU           True                          5   
11922       Tacoma, WA          False                          0   
11923      Seattle, WA          False                          1   

      ratingBusinessoutlook   ratingCEO  ...  rating Overall  \
0                  POSITIVE     APPROVE  ...               4   
1                  POSITIVE     APPROVE  ...               5   
2                       NaN     APPROVE  ...               2   
3                  NEGATIVE  DISAPPROVE  ...               2   
5                  NEGATIVE  NO_OPINION  ...               3   
...                     ...         ...  ...             ...   
11919               NEUTRAL     APPROVE  ...               4   
11920                   NaN         NaN  ...               4   
11921               NEUTRAL  NO_OPINION  ...               5   
11922                   NaN         NaN  ...               5   
11923                   NaN         NaN  ...               2   

       ratingSeniorleadership  rating worklifebalance  countHelpful  \
0                           4                       5             0   
1                           5                       5             0   
2                           2                       3             0   
3                           1                       4             0   
5                           2                       2             0   
...                       ...                     ...           ...   
11919                       4                       5             1   
11920                       0                       0             1   
11921                       2                       5             0   
11922                       0                       0            

### Splitting the location column into city and state and drop location column (Which can be used for visualization in PowerBI/Tableau)

In [83]:
reviews_raw_copy['location'] = reviews_raw_copy['location'].str.split(',')
reviews_raw_copy['city'] = reviews_raw_copy['location'].str[0].str.strip()
reviews_raw_copy['state'] = reviews_raw_copy['location'].str[1].str.strip()

In [85]:
reviews_raw_copy.drop(columns=['location'],inplace=True)

In [86]:
reviews_raw_copy

Employeer ID   Employeer Name Employement Status  \
0           5224839    JPMorganChase            REGULAR   
1           5224839    JPMorganChase            REGULAR   
2           5224839    JPMorganChase            REGULAR   
3           5224839    JPMorganChase            REGULAR   
5           5224839    JPMorganChase            REGULAR   
...             ...              ...                ...   
11919           683  United Airlines            REGULAR   
11920           683  United Airlines            REGULAR   
11921           683  United Airlines          PART_TIME   
11922           683  United Airlines            REGULAR   
11923           683  United Airlines            REGULAR   

                                          job title  length of employement  \
0                             Private Client Banker                      9   
1                                  Senior Associate                      6   
2                                   Personal Banker                      4   
3                             Software Engineer III                      2   
5                                   Process Manager                     20   
...                                             ...                    ...   
11919  Reservation Sales and Service Representative                      1   
11920                              Flight Attendant                      0   
11921                         Ramp Service Employee                      4   
11922                            Operations Manager                      0   
11923                              Flight Attendant                      0   

       isCurrent Job  ratingCarrerOpportunities ratingBusinessoutlook  \
0               True                          4              POSITIVE   
1               True                          5              POSITIVE   
2               True                          2                   NaN   
3               True                          2              NEGATIVE   
5               True                          3              NEGATIVE   
...              ...                        ...                   ...   
11919          False                          4               NEUTRAL   
11920           True                          0                   NaN   
11921           True                          5               NEUTRAL   
11922          False                          0                   NaN   
11923          False                          1                   NaN   

        ratingCEO  ratingCompensationand Benefits  ...  \
0         APPROVE                               3  ...   
1         APPROVE                               5  ...   
2         APPROVE                               4  ...   
3      DISAPPROVE                               4  ...   
5      NO_OPINION                               2  ...   
...           ...                             ...  ...   
11919     APPROVE                               5  ...   
11920         NaN                               0  ...   
11921  NO_OPINION                               5  ...   
11922         NaN                               0  ...   
11923         NaN                               2  ...   

       rating worklifebalance  countHelpful  countNothelpful  \
0                           5             0                0   
1                           5             0                0   
2                           3             0                0   
3                           4             0                0   
5                           2             0                0   
...                       ...           ...              ...   
11919                       5             1                0   
11920                       0             1                0   
11921                       5             0                0   
11922                       0             0                0   
11923                       3             0                0   

                  

### Saving the changes to new CSV File

In [87]:
reviews_raw_copy.to_csv("C:\\Users\\saite\\OneDrive - University of Central Missouri\\Documents\\MS in CS\\Data Analyst\\Projects\\Employee Rentention Analysis\\Datasets\\Glassdoor Reviews Cleaned and Processed.csv", index=False)